src/matching.rs

```diff
@@ -284,6 +284,20 @@ impl<T> DeclarationBlock<T> {
     }
 }
 
+bitflags! {
+    #[doc = "Flags set on elements during the matching process."]
+    flags ElementFlags: u8 {
+        #[doc = "When a child is added or removed from this element, all the children must be"]
+        #[doc = "restyled, because they may match selectors such as :last-of-type."]
+        const HAS_SLOW_SELECTOR = 0x01,
+        #[doc = "When a child is added or removed from this element, any later children must be"]
+        #[doc = "restyled, because they may match selectors such as :nth-child."]
+        const HAS_SLOW_SELECTOR_LATER_SIBLINGS = 0x02,
```

Comment:
> Please add an exhaustive list of which pseudo-classes set which flags.

Okay, will do.

> Is it all child-indexed and type-child-indexed pseudos?

Yes.

> If there’s a flag for later siblings, why not one for earlier siblings with `:nth-last-child()`?

There could be. Gecko doesn't have one, possibly because it was not a big enough win to warrant an additional flag. (It only helps when there is an `:nth-last-child` selector *and* siblings are inserted at the start or middle of the container—both of which are probably quite rare compared to the other cases here.)